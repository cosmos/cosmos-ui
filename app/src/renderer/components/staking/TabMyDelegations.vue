<template lang="pug">
  div
    div(v-if="yourValidators.length > 0")
      h3.tab-header
        | Active Delegations
        |
        i.material-icons.info-button(v-tooltip.top="bondInfo") info_outline
      table-validators(:validators="yourValidators")

    tm-data-msg(v-if="yourValidators.length < 1", icon="info_outline")
      div(slot="title") No Active Delegations
      div(slot="subtitle") Looks like you haven't delegated any {{ this.bondingDenom }}s yet. Head over to the #[router-link(:to="{name: 'Validators'}") validator list] to make your first delegation!

    .check-out-message(v-if="yourValidators.length > 0")
      | Check out
      |
      router-link(:to="{name: 'Validators'}") the validator list
      |
      | to spread some of your Atoms around.

    div(v-if="undelegatedValidators.length")
      h3.tab-header
        | Inactive Delegations
        |
        i.material-icons.info-button(v-tooltip.top="unbondInfo") info_outline
      table-validators(:validators="undelegatedValidators")
</template>

<script>
import { mapGetters } from "vuex"
import { TmDataMsg } from "@tendermint/ui"
import TableValidators from "staking/TableValidators"

export default {
  name: `tab-my-delegations`,
  components: { TableValidators, TmDataMsg },
  data: () => ({
    bondInfo: `Validators you are currently bonded to`,
    unbondInfo: `Your bonded validators in unbonding process`
  }),
  computed: {
    ...mapGetters([
      `delegates`,
      `delegation`,
      `committedDelegations`,
      `bondingDenom`
    ]),
    undelegatedValidators(
      { delegates: { delegates }, delegation: { unbondingDelegations } } = this
    ) {
      return delegates.filter(
        ({ operator_address }) => operator_address in unbondingDelegations
      )
    },
    yourValidators({ committedDelegations, delegates: { delegates } } = this) {
      return delegates.filter(
        ({ operator_address }) => operator_address in committedDelegations
      )
    }
  }
}
</script>
<style lang="stylus">
@require '~variables'

.tab-header
  color var(--dim)
  font-size 14px
  font-weight 500
  margin 1rem 1rem 0 2rem

.info-button
  color var(--link)

.check-out-message
  background var(--app-fg)
  border 1px solid var(--bc-dim)
  border-radius 0.25rem
  font-size sm
  margin-bottom 4rem
  margin-left 2rem
  padding 0.5rem
  text-align center
</style>
