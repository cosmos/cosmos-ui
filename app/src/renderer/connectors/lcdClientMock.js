"use strict"
const b32 = require("../scripts/b32.js")
const { getHeight } = require("./rpcWrapperMock.js")

const botAddress = "cosmosaccaddr1p6zajjw6xged056andyhn62lm7axwzyspkzjq0"
const addresses = [
  "cosmosaccaddr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctpesxxn9",
  "cosmosaccaddr1pxdf0lvq5jvl9uxznklgc5gxuwzpdy5ynem546",
  botAddress
]
const validators = [
  "cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctqzh8yqw",
  "cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctplpn3au",
  "cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctgurrg7n"
]
let state = {
  keys: [
    {
      name: "default",
      password: "1234567890",
      address: addresses[0]
    }
  ],
  accounts: {
    [addresses[0]]: {
      coins: [
        {
          denom: "mycoin",
          amount: "1000"
        },
        {
          denom: "fermion",
          amount: "2300"
        },
        {
          denom: "steak",
          amount: "1000"
        }
      ],
      sequence: "1",
      account_number: "1"
    }
  },
  nonces: { [addresses[0]]: 0 },
  txs: [
    {
      tx: {
        value: {
          msg: [
            {
              value: {
                inputs: [
                  {
                    coins: [
                      {
                        denom: "jbcoins",
                        amount: "1234"
                      }
                    ],
                    address: makeHash()
                  }
                ],
                outputs: [
                  {
                    coins: [
                      {
                        denom: "jbcoins",
                        amount: "1234"
                      }
                    ],
                    address: addresses[0]
                  }
                ]
              }
            }
          ]
        }
      },
      hash: "999ADECC2DE8C3AC2FD4F45E5E1081747BBE504A",
      height: 1
    },
    {
      tx: {
        value: {
          msg: [
            {
              value: {
                inputs: [
                  {
                    coins: [
                      {
                        denom: "fabocoins",
                        amount: "1234"
                      }
                    ],
                    address: addresses[0]
                  }
                ],
                outputs: [
                  {
                    coins: [
                      {
                        denom: "fabocoins",
                        amount: "1234"
                      }
                    ],
                    address: makeHash()
                  }
                ]
              }
            }
          ]
        }
      },
      hash: "A7C6DE5CA923AF08E6088F1348047F16BABB9F48",
      height: 150
    }
  ],
  stake: {
    [addresses[0]]: {
      [validators[0]]: {
        delegator_addr: addresses[0],
        validator_addr: validators[0],
        shares: "130",
        height: 123
      }
    }
  },
  candidates: [
    {
      owner: validators[0],
      pub_key: {
        type: "AC26791624DE60",
        data: "t3zVnKU42WNH+NtYFcstZRLFVULWV8VagoP0HwW43Pk="
      },
      revoked: false,
      tokens: "14",
      delegator_shares: "14",
      description: {
        description: "Mr Mounty",
        moniker: "mr_mounty",
        country: "Canada"
      }
    },
    {
      owner: validators[1],
      pub_key: {
        type: "AC26791624DE60",
        data: "9M4oaDArXKVU5ffqjq2TkynTCMJlyLzpzZLNjHtqM+w="
      },
      tokens: "0",
      delegator_shares: "0",
      description: {
        description: "Good Guy Greg",
        moniker: "good_greg",
        country: "USA"
      }
    },
    {
      owner: validators[2],
      pub_key: {
        type: "AC26791624DE60",
        data: "dlN5SLqeT3LT9WsUK5iuVq1eLQV2Q1JQAuyN0VwSWK0="
      },
      tokens: "19",
      delegator_shares: "19",
      description: {
        description: "Herr Schmidt",
        moniker: "herr_schmidt_revoked",
        country: "DE"
      },
      revoked: true
    }
  ],
  sendHeight: 2
}

module.exports = {
  async lcdConnected() {
    return true
  },

  // keys
  async generateSeed() {
    return "grace admit inherit female grant pledge shine inquiry pencil acid capable damage elegant voice aunt abandon grace admit inherit female grant pledge shine inquiry"
  },
  async storeKey({ name, password, seed }) {
    let key = {
      name,
      password,
      address: makeHash()
    }
    state.keys.push(key)
    return { name, password, seed, address: key.address }
  },
  async listKeys() {
    return state.keys.map(k => ({
      name: k.name,
      address: k.address
    }))
  },
  async getKey(name) {
    return state.keys.find(k => k.name === name)
  },
  async updateKey(account, { name, old_password, new_password }) {
    // eslint-disable-line camelcase
    let key = state.keys.find(k => k.name === name)
    if (key.password !== old_password) {
      // eslint-disable-line camelcase
      throw new Error("Passwords do not match")
    }
    key.password = new_password // eslint-disable-line camelcase
  },
  // axios handles DELETE requests different then other requests, we have to but the body in a config object with the prop data
  async deleteKey(account, { name, password }) {
    let key = state.keys.find(k => k.name === name)
    if (key.password !== password) {
      throw new Error("Passwords do not match")
    }
    state.keys = state.keys.filter(k => k.name !== name)
  },

  // coins
  async queryAccount(address) {
    return state.accounts[address]
  },
  async txs(address) {
    return state.txs.filter(tx => {
      return (
        tx.tx.value.msg[0].value.inputs.find(
          input => input.address === address
        ) ||
        tx.tx.value.msg[0].value.outputs.find(
          output => output.address === address
        )
      )
    })
  },
  async tx() {
    return {}
  },
  async send(to, req) {
    let fromKey = state.keys.find(a => a.name === req.name)
    if (!fromKey)
      throw Error(
        "Key you want to send from does not exist in the lcd connection mock"
      )
    return send(to, fromKey.address, req)
  },
  ibcSend(to, req) {
    // XXX ignores chainId, treated as normal send
    to = to.split("/")[1]
    return module.exports.send(to, req)
  },

  // staking
  async updateDelegations({ name, sequence, delegations, begin_unbondings }) {
    let results = []

    let fromKey = state.keys.find(a => a.name === name)
    let fromAccount = state.accounts[fromKey.address]
    if (fromAccount == null) {
      results.push(txResult(1, "Nonexistent account"))
      return results
    }

    // check nonce
    if (parseInt(fromAccount.sequence) !== parseInt(sequence)) {
      results.push(
        txResult(
          2,
          `Expected sequence "${fromAccount.sequence}", got "${sequence}"`
        )
      )
      return results
    }

    for (let tx of delegations) {
      let { denom } = tx.delegation
      let amount = parseInt(tx.delegation.amount)
      if (amount < 0) {
        results.push(txResult(1, "Amount cannot be negative"))
        return results
      }
      if (fromAccount.coins.find(c => c.denom === denom).amount < amount) {
        results.push(txResult(1, "Not enough coins in your account"))
        return results
      }

      // update sender account
      incrementSequence(fromAccount)
      fromAccount.coins.find(c => c.denom === denom).amount -= amount

      // update stake
      let delegator = state.stake[fromKey.address]
      if (!delegator) {
        state.stake[fromKey.address] = {}
        delegator = state.stake[fromKey.address]
      }
      let delegation = delegator[tx.validator_addr]
      if (!delegation) {
        delegation = {
          delegator_addr: fromKey.address,
          validator_addr: tx.validator_addr,
          shares: "0",
          height: 0
        }
        delegator[tx.validator_addr] = delegation
      }
      let shares = parseInt(delegation.shares)
      delegation.shares = (shares + amount).toString()

      let candidate = state.candidates.find(c => c.owner === tx.validator_addr)
      candidate.tokens = (parseInt(candidate.tokens) + amount).toString()
      candidate.delegator_shares = (
        parseInt(candidate.delegator_shares) + amount
      ).toString()

      results.push(txResult(0))
    }

    for (let tx of begin_unbondings) {
      incrementSequence(fromAccount)

      let amount = parseInt(tx.shares)

      // update sender balance
      let coinBalance = fromAccount.coins.find(c => c.denom === "steak")
      coinBalance.amount = String(parseInt(coinBalance) + amount)

      // update stake
      let delegator = state.stake[fromKey.address]
      if (!delegator) {
        results.push(txResult(2, "Nonexistent delegator"))
        return results
      }
      let delegation = delegator[tx.validator_addr]
      if (!delegation) {
        results.push(txResult(2, "Nonexistent delegation"))
        return results
      }
      let shares = parseInt(delegation.shares)
      delegation.shares = (+shares - amount).toString()

      let candidate = state.candidates.find(c => c.owner === tx.validator_addr)
      shares = parseInt(candidate.tokens)
      candidate.tokens = (+shares - amount).toString()

      results.push(txResult(0))
    }

    return results
  },
  async queryDelegation(delegatorAddress, validatorAddress) {
    let delegator = state.stake[delegatorAddress]
    if (!delegator) return
    return delegator[validatorAddress]
  },
  async candidates() {
    return state.candidates
  },
  async getValidators() {
    return {
      block_height: 1,
      validators: state.candidates
    }
  },
  // exports to be used in tests
  state,
  addresses,
  validators
}

function makeHash() {
  var text = ""
  var possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"

  for (var i = 0; i < 40; i++) {
    text += possible.charAt(Math.floor(Math.random() * possible.length))
  }
  return b32.encode(text)
}

function send(to, from, req) {
  let fromAccount = state.accounts[from]
  if (fromAccount == null) {
    return txResult(1, "Nonexistent account")
  }

  for (let { denom, amount } of req.amount) {
    if (parseInt(amount) < 0) {
      return txResult(1, "Amount cannot be negative")
    }
    if (
      fromAccount.coins.find(c => c.denom === denom).amount < parseInt(amount)
    ) {
      return txResult(1, "Not enough coins in your account")
    }
  }

  // check/update nonce
  if (parseInt(fromAccount.sequence) !== parseInt(req.sequence)) {
    return txResult(
      2,
      `Expected sequence "${fromAccount.sequence}", got "${req.sequence}"`
    )
  }
  incrementSequence(fromAccount)

  // update sender balances
  for (let { denom, amount } of req.amount) {
    let senderBalance = fromAccount.coins.find(c => c.denom === denom)
    senderBalance.amount = String(
      parseInt(senderBalance.amount) - parseInt(amount)
    )
  }

  // update receiver balances
  let receiverAccount = state.accounts[to]
  if (!receiverAccount) {
    receiverAccount = state.accounts[to] = {
      coins: [],
      sequence: "0"
    }
  }
  for (let { denom, amount } of req.amount) {
    let receiverBalance = receiverAccount.coins.find(c => c.denom === denom)
    if (!receiverBalance) {
      receiverBalance = { amount: "0", denom }
      receiverAccount.coins.push(receiverBalance)
    }
    receiverBalance.amount = String(
      parseInt(receiverBalance.amount) + parseInt(amount)
    )
  }

  // log tx
  state.txs.push({
    tx: {
      value: {
        msg: [
          {
            value: {
              inputs: [
                {
                  coins: req.amount,
                  address: from
                }
              ],
              outputs: [
                {
                  coins: req.amount,
                  address: to
                }
              ]
            }
          }
        ]
      }
    },
    hash: makeHash(),
    height: getHeight() + (from === botAddress ? 1 : 0),
    time: Date.now()
  })

  // if receiver is bot address, send money back
  if (to === botAddress) {
    send(from, botAddress, {
      amount: req.amount,
      sequence: state.accounts[botAddress].sequence
    })
  }

  return txResult(0)
}

// function delegate (sender, { pub_key: { data: pubKey }, amount: delegation }) {
//   let delegate = state.delegates.find(d => d.pub_key.data === pubKey)
//   if (!delegate) {
//     return txResult(1, 'Delegator does not exist')
//   }
//
//   let fermions = state.accounts[sender].coins.find(c => c.denom === 'fermion')
//   if (!fermions) {
//     state.accounts[sender].coins.push({ denom: 'fermion', amount: 0 })
//     fermions = state.accounts[sender].coins.find(c => c.denom === 'fermion')
//   }
//   if (fermions.amount < delegation.amount) {
//     return txResult(1, 'Not enought fermions to stake')
//   }
//
//   // execute
//   fermions.amount -= delegation.amount
//   delegate.voting_power += delegation.amount
//   state.stake[sender][pubKey] = state.stake[sender][pubKey] || {
//     PubKey: { data: pubKey },
//     Shares: 0
//   }
//   state.stake[sender][pubKey].Shares += delegation.amount
//
//   return txResult()
// }

function txResult(code = 0, message = "") {
  return {
    check_tx: {
      code: code,
      data: "",
      log: message,
      gas: "0",
      fee: "0"
    },
    deliver_tx: {
      code: 0,
      data: "",
      log: "",
      tags: []
    },
    hash: "999ADECC2DE8C3AC2FD4F45E5E1081747BBE504A",
    height: 0
  }
}

function incrementSequence(account) {
  account.sequence = (parseInt(account.sequence) + 1).toString()
}
