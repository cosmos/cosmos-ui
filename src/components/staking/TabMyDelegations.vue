<template>
  <div>
    <CardSignInRequired v-if="!session.signedIn" />
    <TmDataConnecting
      v-else-if="!connected && $apollo.queries.validators.loading"
    />
    <TmDataLoading v-else-if="$apollo.queries.validators.loading" />
    <TmDataMsg
      v-else-if="!$apollo.queries.validators.loading && validators.length === 0"
      icon="sentiment_dissatisfied"
    >
      <div slot="title">No Active Delegations</div>
      <div slot="subtitle">
        Looks like you haven't delegated any {{ bondDenom | viewDenom }}s yet.
        Head over to the
        <router-link :to="{ name: 'Validators' }">validator list</router-link>to
        make your first delegation!
      </div>
    </TmDataMsg>
    <div v-else>
      <TableValidators
        :validators="validators"
        show-on-mobile="my_delegations"
      />
    </div>
    <div v-if="delegation.loaded && pendingUndelegations.length > 0">
      <h3 class="tab-header transactions">Pending Undelegations</h3>
      <div class="unbonding-transactions">
        <template>
          <TransactionList
            :transactions="pendingUndelegations"
            :address="session.address"
            :validators="yourValidatorsAddressMap"
          />
          <br />
        </template>
      </div>
    </div>
  </div>
</template>

<script>
import { mapState, mapGetters } from "vuex"
import { viewDenom } from "scripts/num"
import { isPendingUndelegation } from "scripts/transaction-utils"
import { SomeValidators, AllValidatorsResult } from "src/gql"

import TmDataMsg from "common/TmDataMsg"
import CardSignInRequired from "common/CardSignInRequired"
import TmDataLoading from "common/TmDataLoading"
import TableValidators from "staking/TableValidators"
import TmDataConnecting from "common/TmDataConnecting"
import TransactionList from "transactions/TransactionList"

export default {
  name: `tab-my-delegations`,
  components: {
    TableValidators,
    TmDataMsg,
    TmDataConnecting,
    TmDataLoading,
    CardSignInRequired,
    TransactionList
  },
  filters: {
    viewDenom
  },
  data: () => ({
    validators: []
  }),
  computed: {
    ...mapState([`delegates`, `delegation`, `session`]),
    ...mapGetters([
      `committedDelegations`,
      `bondDenom`,
      `connected`,
      `flatOrderedTransactionList`
    ]),
    yourValidatorsAddressMap() {
      const names = {}
      this.validators.forEach(item => {
        names[item.operator_address] = item
      })
      return names
    },
    pendingUndelegations() {
      return this.flatOrderedTransactionList.filter(isPendingUndelegation)
    }
  },
  watch: {
    "session.signedIn": function() {
      this.loadStakingTxs()
    }
  },
  async created() {
    this.loadStakingTxs()
  },
  methods: {
    async loadStakingTxs() {
      if (this.session.signedIn) {
        await this.$store.dispatch(`getAllTxs`)
      }
    }
  },
  apollo: {
    validators: {
      query: SomeValidators,
      variables() {
        return {
          addressList: Object.keys(this.committedDelegations)
        }
      },
      update: AllValidatorsResult
    }
  }
}
</script>
<style>
.tab-header {
  color: var(--dim);
  font-size: 14px;
  font-weight: 500;
  margin: 3rem 0.5rem 0.5rem;
}

.info-button {
  color: var(--link);
}

.unbonding-transactions .tm-li-tx::before {
  position: absolute;
  width: 2rem;
  text-align: right;
  color: var(--dim);
  left: 0;
}
</style>
