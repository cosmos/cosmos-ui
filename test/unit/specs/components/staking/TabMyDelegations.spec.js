import setup from "../../../helpers/vuex-setup"
import lcdClientMock from "renderer/connectors/lcdClientMock.js"
import TabMyDelegations from "renderer/components/staking/TabMyDelegations"

const delegates = lcdClientMock.candidates

describe(`Component: TabMyDelegations`, () => {
  let { mount } = setup()

  it(`should show committed validators`, () => {
    let instance = mount(TabMyDelegations, {
      getters: {
        committedDelegations: () => ({
          [delegates[0].operator_address]: 42
        }),
        delegates: () => ({
          delegates
        }),
        delegation: () => ({
          unbondingDelegations: {
            [delegates[1].operator_address]: 1,
            [delegates[2].operator_address]: 2
          }
        }),
        bondingDenom: () => `stake`
      }
    })

    expect(instance.wrapper.html()).toContain(`Active Delegations`)
    expect(instance.wrapper.vm.$el).toMatchSnapshot()
  })

  it(`should show unbonding validators`, () => {
    let instance = mount(TabMyDelegations, {
      getters: {
        committedDelegations: () => ({}),
        delegates: () => ({
          delegates
        }),
        delegation: () => ({
          unbondingDelegations: {
            [delegates[1].operator_address]: 1,
            [delegates[2].operator_address]: 2
          }
        }),
        bondingDenom: () => `stake`
      }
    })

    expect(instance.wrapper.html()).toContain(`Inactive Delegations`)
    expect(instance.wrapper.vm.$el).toMatchSnapshot()
  })

  it(`should show a message if not staked yet to any validator`, () => {
    let instance = mount(TabMyDelegations, {
      getters: {
        committedDelegations: () => ({}),
        delegates: () => ({
          delegates
        }),
        delegation: () => ({
          unbondingDelegations: {}
        }),
        bondingDenom: () => `stake`
      }
    })

    expect(instance.wrapper.html()).toContain(`No Active Delegations`)
    expect(instance.wrapper.vm.$el).toMatchSnapshot()
  })

  it(`undelegatedValidators`, () => {
    expect(
      TabMyDelegations.computed.undelegatedValidators({
        delegation: {
          unbondingDelegations: {
            [delegates[1].operator_address]: 1,
            [delegates[2].operator_address]: 2
          }
        },
        delegates: { delegates }
      })
    ).toEqual([delegates[1], delegates[2]])
  })

  it(`yourValidators`, () => {
    expect(
      TabMyDelegations.computed.yourValidators({
        committedDelegations: {
          [delegates[0].operator_address]: 1,
          [delegates[2].operator_address]: 2
        },
        delegates: { delegates }
      })
    ).toEqual([delegates[0], delegates[2]])
  })
})
