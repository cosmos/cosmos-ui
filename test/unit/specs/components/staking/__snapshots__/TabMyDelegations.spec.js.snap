// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: TabMyDelegations should show a message if not staked yet to any validator 1`] = `
<div>
  <!---->
  <div
    class="tm-data-msg"
  >
    <div
      class="tm-data-msg__icon"
    >
      <i
        class="material-icons"
      >
        info_outline
      </i>
    </div>
    <div
      class="tm-data-msg__text"
    >
      <div
        class="tm-data-msg__title"
      >
        <div>
          No Active Delegations
        </div>
      </div>
      <div
        class="tm-data-msg__subtitle"
      >
        <div>
          Looks like you haven't delegated any stakes yet. Head over to the 
          <a
            class=""
            href="#/staking/Validators"
          >
            validator list
          </a>
           to make your first delegation!
        </div>
      </div>
    </div>
  </div>
  <!---->
  <!---->
</div>
`;

exports[`Component: TabMyDelegations should show committed validators 1`] = `
<div>
  <div>
    <h3
      class="tab-header"
    >
      Active Delegations

      <i
        class="material-icons info-button"
      >
        info_outline
      </i>
    </h3>
    <div>
      <table>
        <thead>
          <tr
            class="panel-sort-container"
          >
            <th
              class="sort-by name"
            >
              <a
                class="sort-by-link"
              >
                Moniker
              </a>
            </th>
            <th
              class="sort-by your-votes"
            >
              <a
                class="sort-by-link"
              >
                Delegated stake
              </a>
            </th>
            <th
              class="sort-by your-rewards"
            >
              <a
                class="sort-by-link"
              >
                Rewards
              </a>
            </th>
            <th
              class="sort-by percent_of_vote"
            >
              <a
                class="sort-by-link"
              >
                Voting Power
              </a>
            </th>
            <th
              class="sort-by uptime"
            >
              <a
                class="sort-by-link"
              >
                Uptime
              </a>
            </th>
            <th
              class="sort-by commission"
            >
              <a
                class="sort-by-link"
              >
                Commission
              </a>
            </th>
            <th
              class="sort-by slashes"
            >
              <a
                class="sort-by-link"
              >
                Slashes
              </a>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="li-validator"
          >
            <td
              class="li-validator__moniker-container"
            >
              <img
                class="li-validator__avatar no-img"
                height="48"
                src="~assets/images/validator-icon.svg"
                width="48"
              />
              <div
                class="li-validator__name-container"
              >
                <span
                  class="validator-profile__status yellow"
                />
                <a
                  class="li-validator__moniker"
                  href="#/staking/validators/cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctqzh8yqw"
                >
                  mr_mounty
                </a>
                <div
                  class="short-bech32 li-validator__address"
                >
                  <div
                    class="address"
                    id="address"
                  >
                    cosmosvaladdr…8yqw
                  </div>
                  <div
                    class="copied"
                  >
                    <i
                      class="material-icons"
                    >
                      check
                    </i>
                    <span>
                      Copied
                    </span>
                  </div>
                </div>
              </div>
            </td>
            <td
              class="li-validator__delegated-steak"
            >
              0.0000…
            </td>
            <td
              class="li-validator__rewards"
            >
              n/a
            </td>
            <td
              class="li-validator__voting-power"
            >
              0.00%
            </td>
            <td
              class="li-validator__uptime"
            >
              n/a
            </td>
            <td
              class="li-validator__commission"
            >
              0.00%
            </td>
            <td
              class="li-validator__slashes"
            >
              n/a
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <!---->
  <div
    class="check-out-message"
  >
    Check out

    <a
      class=""
      href="#/staking/Validators"
    >
      the validator list
    </a>
    
to spread some of your Atoms around.
  </div>
  <div>
    <h3
      class="tab-header"
    >
      Inactive Delegations

      <i
        class="material-icons info-button"
      >
        info_outline
      </i>
    </h3>
    <div>
      <table>
        <thead>
          <tr
            class="panel-sort-container"
          >
            <th
              class="sort-by name"
            >
              <a
                class="sort-by-link"
              >
                Moniker
              </a>
            </th>
            <th
              class="sort-by your-votes"
            >
              <a
                class="sort-by-link"
              >
                Delegated stake
              </a>
            </th>
            <th
              class="sort-by your-rewards"
            >
              <a
                class="sort-by-link"
              >
                Rewards
              </a>
            </th>
            <th
              class="sort-by percent_of_vote"
            >
              <a
                class="sort-by-link"
              >
                Voting Power
              </a>
            </th>
            <th
              class="sort-by uptime"
            >
              <a
                class="sort-by-link"
              >
                Uptime
              </a>
            </th>
            <th
              class="sort-by commission"
            >
              <a
                class="sort-by-link"
              >
                Commission
              </a>
            </th>
            <th
              class="sort-by slashes"
            >
              <a
                class="sort-by-link"
              >
                Slashes
              </a>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="li-validator"
          >
            <td
              class="li-validator__moniker-container"
            >
              <img
                class="li-validator__avatar no-img"
                height="48"
                src="~assets/images/validator-icon.svg"
                width="48"
              />
              <div
                class="li-validator__name-container"
              >
                <span
                  class="validator-profile__status yellow"
                />
                <a
                  class="li-validator__moniker"
                  href="#/staking/validators/cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctplpn3au"
                >
                  good_greg
                </a>
                <div
                  class="short-bech32 li-validator__address"
                >
                  <div
                    class="address"
                    id="address"
                  >
                    cosmosvaladdr…n3au
                  </div>
                  <div
                    class="copied"
                  >
                    <i
                      class="material-icons"
                    >
                      check
                    </i>
                    <span>
                      Copied
                    </span>
                  </div>
                </div>
              </div>
            </td>
            <td
              class="li-validator__delegated-steak"
            >
              0.0000…
            </td>
            <td
              class="li-validator__rewards"
            >
              n/a
            </td>
            <td
              class="li-validator__voting-power"
            >
              0.00%
            </td>
            <td
              class="li-validator__uptime"
            >
              n/a
            </td>
            <td
              class="li-validator__commission"
            >
              0.00%
            </td>
            <td
              class="li-validator__slashes"
            >
              n/a
            </td>
          </tr>
          <tr
            class="li-validator"
          >
            <td
              class="li-validator__moniker-container"
            >
              <img
                class="li-validator__avatar no-img"
                height="48"
                src="~assets/images/validator-icon.svg"
                width="48"
              />
              <div
                class="li-validator__name-container"
              >
                <span
                  class="validator-profile__status red"
                />
                <a
                  class="li-validator__moniker"
                  href="#/staking/validators/cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctgurrg7n"
                >
                  herr_schmidt_revoked
                </a>
                <div
                  class="short-bech32 li-validator__address"
                >
                  <div
                    class="address"
                    id="address"
                  >
                    cosmosvaladdr…rg7n
                  </div>
                  <div
                    class="copied"
                  >
                    <i
                      class="material-icons"
                    >
                      check
                    </i>
                    <span>
                      Copied
                    </span>
                  </div>
                </div>
              </div>
            </td>
            <td
              class="li-validator__delegated-steak"
            >
              0.0000…
            </td>
            <td
              class="li-validator__rewards"
            >
              n/a
            </td>
            <td
              class="li-validator__voting-power"
            >
              0.00%
            </td>
            <td
              class="li-validator__uptime"
            >
              n/a
            </td>
            <td
              class="li-validator__commission"
            >
              0.00%
            </td>
            <td
              class="li-validator__slashes"
            >
              n/a
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`Component: TabMyDelegations should show unbonding validators 1`] = `
<div>
  <!---->
  <div
    class="tm-data-msg"
  >
    <div
      class="tm-data-msg__icon"
    >
      <i
        class="material-icons"
      >
        info_outline
      </i>
    </div>
    <div
      class="tm-data-msg__text"
    >
      <div
        class="tm-data-msg__title"
      >
        <div>
          No Active Delegations
        </div>
      </div>
      <div
        class="tm-data-msg__subtitle"
      >
        <div>
          Looks like you haven't delegated any stakes yet. Head over to the 
          <a
            class=""
            href="#/staking/Validators"
          >
            validator list
          </a>
           to make your first delegation!
        </div>
      </div>
    </div>
  </div>
  <!---->
  <div>
    <h3
      class="tab-header"
    >
      Inactive Delegations

      <i
        class="material-icons info-button"
      >
        info_outline
      </i>
    </h3>
    <div>
      <table>
        <thead>
          <tr
            class="panel-sort-container"
          >
            <th
              class="sort-by name"
            >
              <a
                class="sort-by-link"
              >
                Moniker
              </a>
            </th>
            <th
              class="sort-by your-votes"
            >
              <a
                class="sort-by-link"
              >
                Delegated stake
              </a>
            </th>
            <th
              class="sort-by your-rewards"
            >
              <a
                class="sort-by-link"
              >
                Rewards
              </a>
            </th>
            <th
              class="sort-by percent_of_vote"
            >
              <a
                class="sort-by-link"
              >
                Voting Power
              </a>
            </th>
            <th
              class="sort-by uptime"
            >
              <a
                class="sort-by-link"
              >
                Uptime
              </a>
            </th>
            <th
              class="sort-by commission"
            >
              <a
                class="sort-by-link"
              >
                Commission
              </a>
            </th>
            <th
              class="sort-by slashes"
            >
              <a
                class="sort-by-link"
              >
                Slashes
              </a>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="li-validator"
          >
            <td
              class="li-validator__moniker-container"
            >
              <img
                class="li-validator__avatar no-img"
                height="48"
                src="~assets/images/validator-icon.svg"
                width="48"
              />
              <div
                class="li-validator__name-container"
              >
                <span
                  class="validator-profile__status yellow"
                />
                <a
                  class="li-validator__moniker"
                  href="#/staking/validators/cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctplpn3au"
                >
                  good_greg
                </a>
                <div
                  class="short-bech32 li-validator__address"
                >
                  <div
                    class="address"
                    id="address"
                  >
                    cosmosvaladdr…n3au
                  </div>
                  <div
                    class="copied"
                  >
                    <i
                      class="material-icons"
                    >
                      check
                    </i>
                    <span>
                      Copied
                    </span>
                  </div>
                </div>
              </div>
            </td>
            <td
              class="li-validator__delegated-steak"
            >
              0.0000…
            </td>
            <td
              class="li-validator__rewards"
            >
              n/a
            </td>
            <td
              class="li-validator__voting-power"
            >
              0.00%
            </td>
            <td
              class="li-validator__uptime"
            >
              n/a
            </td>
            <td
              class="li-validator__commission"
            >
              0.00%
            </td>
            <td
              class="li-validator__slashes"
            >
              n/a
            </td>
          </tr>
          <tr
            class="li-validator"
          >
            <td
              class="li-validator__moniker-container"
            >
              <img
                class="li-validator__avatar no-img"
                height="48"
                src="~assets/images/validator-icon.svg"
                width="48"
              />
              <div
                class="li-validator__name-container"
              >
                <span
                  class="validator-profile__status red"
                />
                <a
                  class="li-validator__moniker"
                  href="#/staking/validators/cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctgurrg7n"
                >
                  herr_schmidt_revoked
                </a>
                <div
                  class="short-bech32 li-validator__address"
                >
                  <div
                    class="address"
                    id="address"
                  >
                    cosmosvaladdr…rg7n
                  </div>
                  <div
                    class="copied"
                  >
                    <i
                      class="material-icons"
                    >
                      check
                    </i>
                    <span>
                      Copied
                    </span>
                  </div>
                </div>
              </div>
            </td>
            <td
              class="li-validator__delegated-steak"
            >
              0.0000…
            </td>
            <td
              class="li-validator__rewards"
            >
              n/a
            </td>
            <td
              class="li-validator__voting-power"
            >
              0.00%
            </td>
            <td
              class="li-validator__uptime"
            >
              n/a
            </td>
            <td
              class="li-validator__commission"
            >
              0.00%
            </td>
            <td
              class="li-validator__slashes"
            >
              n/a
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
